core/src/main/java/com/google/googlejavaformat/java/java15/Java15InputAstVisitor.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2021 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.googlejavaformat.java.java15;
+
+import com.google.googlejavaformat.OpsBuilder;
+import com.google.googlejavaformat.java.java14.Java14InputAstVisitor;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.Tree;
+import java.util.List;
+
+/** Extends {@link Java14InputAstVisitor} with support for AST nodes that were added for Java 15. */
+public class Java15InputAstVisitor extends Java14InputAstVisitor {
+
+  public Java15InputAstVisitor(OpsBuilder builder, int indentMultiplier) {
+    super(builder, indentMultiplier);
+  }
+
+  @Override
+  protected List<? extends Tree> getPermitsClause(ClassTree node) {
+    return node.getPermitsClause();
+  }
```

Comment:
One more thought here: we're going to need to add more version-specific visitors eventually, especially if there are new `visit*` methods for new AST nodes. However for cases where there's a new getter on an existing node, we're already using reflection in a couple of places, e.g. here: https://github.com/google/google-java-format/blob/bf5ef1c5de94fd623025e24b99ceadf18e83d8a5/core/src/main/java/com/google/googlejavaformat/java/java14/Java14InputAstVisitor.java#L60

I wonder if it'd be simpler to call `getPermitsClause` reflectively and defer adding a new visitor until there are new AST nodes, and to also clean up some of the reflection in `Java14InputAstVisitor` when that happens?

This is fairly subjective and I don't mind making that change when I import it, but let me know if you have any input on that.

Reply:
I didn’t quite understand why these methods were being invoked reflectively in `Java14InputAsVisitor` to begin with.  The code is only compiled in a version that should have it.  Is it to handle the potential case of a preview feature getting changed/dropped in a future Java release?  Regarding `getPermitsClause`, that’s the final API in Java 17, as far as I can tell.  I don’t mind having it invoked reflectively in `Java15InputAstVisitor`, but I’d like to know the reason behind it (sorry, I should have asked earlier).

Or do you mean that we should perhaps drop `Java15InputAstVisitor` altogether and do the reflective call in `JavaInputAstVisitor` instead?  I’m fine with that too.  That’d simplify `pom.xml` and `Formatter`.  I implemented it this way to keep version-specific code out of `JavaInputAstVisitor` as much as possible, but perhaps this minor thing isn’t worth the effort.

Reply:
> Or do you mean that we should perhaps drop Java15InputAstVisitor altogether and do the reflective call in JavaInputAstVisitor instead?

Thanks, yes, I think I'd slightly prefer to drop `Java15InputAstVisitor` and use reflection in either `JavaInputAstVisitor` or `Java14InputAstVisitor`.

`getModule` is invoked reflectively because that method was added to `CompilationUnitTree` in JDK 17. In general the uses of reflection in `Java14InputAstVisitor` should all be for APIs that aren't the same for all versions >= 14 (because they were preview APIs that changed, or they were added in versions after 14).

My tentative plan for supporting future language versions is to introduce new `JavaNInputAstVisitor` classes for major versions that add new AST nodes and `visit*` methods, but to not necessarily add a new `Visitor` class for major versions if the only changes were in getters and it's not too unwieldy to use reflection.